Add RetrofitCallAdapterFactory

bychkovdmitry opened the pull request:
### Description

Introduces RetrofitCallAdapterFactory which allows us to use chat Call as a return type in RetrofitApi

### Checklist

- [X] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [X] PR targets the `develop` branch
- [ ] Changelog updated with client-facing changes
- [X] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [X] Reviewers added

Commit:
Add RetrofitCallAdapterFactory

JcMinarro reviewed (commented):
Can't we totally remove `RetrofitCallMapper` class?
On the other hand, `ChatApi` interface seems useless, it is only implemented by `ChatApiImpl` What do you think about to move `ChatApiImpl` code to `ChatApi` and convert it to class?

Commit:
Remove RetrofitCallMapper class and ChatApi interface

bychkovdmitry commented:
> Can't we totally remove `RetrofitCallMapper` class?
> On the other hand, `ChatApi` interface seems useless, it is only implemented by `ChatApiImpl` What do you think about to move `ChatApiImpl` code to `ChatApi` and convert it to class?

Removed ChatApi interface, removed RetrofitCallMapper. Added possibility to get the original retrofit call from chat call for this case:

`retrofitCdnApi .sendImage(...).originalCall().enqueue(RetroProgressCallback(callback))`

zsmb13 commented on client/src/main/java/io/getstream/chat/android/client/call/RetrofitCall.kt:
```diff
@@ -35,6 +35,8 @@ internal class RetrofitCall<T : Any>(
         call.cancel()
     }
 
+    fun originalCall(): retrofit2.Call<T> = call
```
Nit: could be exposed as just the property itself, not sure we need this separate getter for it.

zsmb13 commented on client/src/main/java/io/getstream/chat/android/client/api/RetrofitCallAdapterFactory.kt:
```diff
@@ -0,0 +1,46 @@
+        return RetrofitCallAdapter<Any>(responseType, chatParser)
+    }
+
+    companion object {
+        @JvmStatic
+        fun create(chatParser: ChatParser) = RetrofitCallAdapterFactory(chatParser)
+    }
```
Nit: we don't need `@JvmStatic`, as we don't have Java code 😄 . We could also drop this factory method, and just use the constructor of the class directly.

zsmb13 commented on client/src/test/java/io/getstream/chat/android/client/api/RetrofitCallAdapterFactoryTests.kt:
```diff
@@ -0,0 +1,68 @@
+import retrofit2.Retrofit
+import java.lang.reflect.Type
+
+internal class RetrofitCallAdapterFactoryTests {
```
Nice coverage of the different cases here 👍

zsmb13 reviewed (dismissed):
Looks great, I have just a couple small comments

zsmb13 committed:
Merge branch 'develop' into retrofit-call-adapter-factory

Commit:
Add RetrofitCallAdapterFactory comments

Commit:
Merge branch 'retrofit-call-adapter-factory' of https://github.com/GetStream/stream-chat-android-client into retrofit-call-adapter-factory
